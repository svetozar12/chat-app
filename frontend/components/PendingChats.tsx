import { AppProps } from "next/dist/shared/lib/router/router";
import axios from "axios";
import React from "react";
import { useCookie } from "next-cookie";
import { AiFillCheckCircle, AiFillCloseCircle } from "react-icons/ai";
function PendingChats({
  _id,
  inviter,
  status,
  reciever,
  setLocalStatus,
  cookie,
  socketRef,
}: AppProps) {
  const cokie = useCookie(cookie);

  const updateInviteStatus = async (word: string) => {
    try {
      setLocalStatus(word);
      const res = await axios.put(`http://localhost:4001/invites`, {
        id: _id,
        status: word,
      });
      setLocalStatus("");
      return true;
    } catch (error) {
      return false;
    }
  };

  const emitFriendRequest = (param: string) => {
    socketRef?.emit("friend_request", {
      reciever,
      inviter,
      status: param,
    });
  };

  return (
    <div>
      {status === "recieved" && (
        <div className="contacts">
          <h1>{inviter}</h1>
          <div className="invite_buttons">
            <button
              onClick={() => {
                updateInviteStatus("accepted");
                emitFriendRequest("accepted");
              }}
              className="accept"
            >
              <AiFillCheckCircle />
            </button>
            <button
              onClick={() => {
                updateInviteStatus("declined");
                emitFriendRequest("declined");
              }}
              className="decline"
            >
              <AiFillCloseCircle />
            </button>
          </div>
        </div>
      )}
    </div>
  );
}

export default PendingChats;
