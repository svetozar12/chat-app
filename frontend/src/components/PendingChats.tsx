import axios from "axios";
import React from "react";
import { useCookie } from "next-cookie";
import { AiFillCheckCircle, AiFillCloseCircle } from "react-icons/ai";
import { Socket } from "socket.io-client";
function PendingChats({
  _id,
  inviter,
  status,
  reciever,
  socketRef,
  setLocalStatus,
  data,
}: {
  _id: string;
  inviter: string;
  status: string;
  reciever: string;
  socketRef: Socket;
  setLocalStatus: any;
  data: string[];
}) {
  const emitFriendRequest = async () => {
    socketRef?.emit("friend_request");
  };

  const updateInviteStatus = async (param: string) => {
    try {
      setLocalStatus(param);
      const res = await axios.put(`http://localhost:4001/invites`, {
        id: _id,
        status: param,
      });
      emitFriendRequest();
      setLocalStatus("");
      return true;
    } catch (error) {
      return false;
    }
  };

  const createChatRoom = async () => {
    try {
      const res = await axios.post("http://localhost:4001/chat-room/messages", {
        user1: inviter,
        user2: reciever,
        sender: inviter,
        message: "",
      });
    } catch (error) {
      console.log(error);
      return false;
    }
  };

  return (
    <div>
      {status === "recieved" && (
        <div className="contacts">
          <h1>{inviter}</h1>
          <div className="invite_buttons">
            <button
              onClick={() => {
                createChatRoom();
                updateInviteStatus("accepted");
              }}
              className="accept"
            >
              <AiFillCheckCircle />
            </button>
            <button
              onClick={() => updateInviteStatus("declined")}
              className="decline"
            >
              <AiFillCloseCircle />
            </button>
          </div>
        </div>
      )}
    </div>
  );
}

export default PendingChats;
