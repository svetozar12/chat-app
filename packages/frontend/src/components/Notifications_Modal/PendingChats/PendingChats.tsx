import React from 'react';
import { AiFillCheckCircle, AiFillCloseCircle } from 'react-icons/ai';
import { css } from '@emotion/css';
import { useCookie } from 'next-cookie';
import { Button, Heading, HStack, Spacer, VStack } from '@chakra-ui/react';
import { Iinvites } from 'pages/[acc]';
import SingleAvatar from 'components/Avatar/SingleAvatar';
// services
import apiHelper from 'services/graphql/apiHelper';
import { connect } from 'react-redux';
import { STATE } from 'services/redux/reducer';
import { IWebSocket } from 'services/redux/reducer/websocket/state';

interface IPendingChats extends Iinvites {
  _id: string;
  inviter: string;
  status: string;
  reciever: string;
  ws: IWebSocket;
}

function PendingChats(props: IPendingChats) {
  const { _id, inviter, reciever, status, ws } = props;
  const cookie = useCookie();
  const id: string = cookie.get('id');
  const inviteId = _id;
  const token: string = cookie.get('token');

  const emitFriendRequest = async () => {
    ws.ws?.emit('friend_request');
  };

  const updateInviteStatus = async (param: string) => {
    try {
      await apiHelper.invite.update(id, inviteId, param, token);
      emitFriendRequest();

      return true;
    } catch (error) {
      return false;
    }
  };

  const createChatRoom = async () => {
    try {
      await apiHelper.chatroom.create({ userId: id, invite_id: _id, user1: inviter, user2: reciever }, token);
      emitFriendRequest();
    } catch (error) {
      return false;
    }
  };

  const isRecieved = status === 'recieved' && inviter !== cookie.get('name');

  const buttons = [
    {
      Svg: AiFillCheckCircle,
      Button,
      props: {
        bg: 'transparent',
        color: 'green.500',
        transition: '0.2s',
        _hover: { opacity: '0.8', transition: '0.2s' },
        onClick: async () => await createChatRoom(),
      },
    },
    {
      Svg: AiFillCloseCircle,
      Button,
      props: {
        bg: 'transparent',
        color: 'red.600',
        transition: '0.2s',
        _hover: { opacity: '0.8', transition: '0.2s' },
        onClick: async () => await updateInviteStatus('declined'),
      },
    },
  ];

  return (
    <HStack w="full">
      {isRecieved && (
        <HStack w="98.5%" m={2} h="20vh">
          <VStack align="center">
            <SingleAvatar width="3rem" height="3rem" />
            <Heading>{inviter}</Heading>
          </VStack>
          <Spacer />
          <HStack>
            {buttons.map((element, index) => {
              const { props, Svg } = element;

              return (
                <Button key={index} {...props}>
                  <Svg
                    className={css`
                      width: 3rem;
                      height: 3rem;
                    `}
                  />
                </Button>
              );
            })}
          </HStack>
        </HStack>
      )}
    </HStack>
  );
}

const mapStateToProps = (state: STATE) => ({
  ws: state.ws,
});

export default connect(mapStateToProps)(PendingChats);
