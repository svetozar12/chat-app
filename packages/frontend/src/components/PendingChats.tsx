import axios from "axios";
import React from "react";
import { AiFillCheckCircle, AiFillCloseCircle } from "react-icons/ai";
import { Socket } from "socket.io-client";
import { requestUrl } from "../utils/hostUrl_requestUrl";
import { Iinvites } from "../pages/[acc]";
interface IPendingChats extends Iinvites {
  _id: string;
  inviter: string;
  status: string;
  reciever: string;
  users: string[];
  socketRef: Socket;
  setLocalStatus: React.Dispatch<React.SetStateAction<string>>;
}

function PendingChats({
  _id,
  inviter,
  status,
  reciever,
  users,
  socketRef,
  setLocalStatus,
}: IPendingChats) {
  const emitFriendRequest = async () => {
    socketRef?.emit("friend_request");
  };

  const updateInviteStatus = async (param: string) => {
    try {
      setLocalStatus(param);
      const res = await axios.put(`${requestUrl}/invites`, {
        id: _id,
        status: param,
      });
      emitFriendRequest();
      setLocalStatus("");

      return true;
    } catch (error) {
      return false;
    }
  };

  const createChatRoom = async () => {
    try {
      setLocalStatus("accepted");
      await axios.put(`${requestUrl}/chat-room`, {
        id: _id,
        userData: users,
      });
      emitFriendRequest();
      setLocalStatus("");
    } catch (error) {
      return false;
    }
  };

  return (
    <div>
      {status === "recieved" && (
        <div className="contacts">
          <h1>{inviter}</h1>
          <div className="invite_buttons">
            <button
              onClick={() => {
                createChatRoom();
              }}
              className="accept"
            >
              <AiFillCheckCircle />
            </button>
            <button
              onClick={() => updateInviteStatus("declined")}
              className="decline"
            >
              <AiFillCloseCircle />
            </button>
          </div>
        </div>
      )}
    </div>
  );
}

export default PendingChats;
