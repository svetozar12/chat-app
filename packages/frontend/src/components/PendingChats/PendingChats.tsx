import axios from "axios";
import React from "react";
import { AiFillCheckCircle, AiFillCloseCircle } from "react-icons/ai";
import { Socket } from "socket.io-client";
import { requestUrl } from "../../utils/hostUrl_requestUrl";
import { Iinvites } from "../../pages/[acc]";
import { FaUserCircle } from "react-icons/fa";
import styled from "@emotion/styled";
import { css } from "@emotion/css";

interface IPendingChats extends Iinvites {
  _id: string;
  inviter: string;
  status: string;
  reciever: string;
  socketRef: Socket;
  setLocalStatus: React.Dispatch<React.SetStateAction<string>>;
}

const Contacts = styled.div`
  border-top: 2px solid rgba(0, 0, 0, 0.3);
  background: var(--main-white);
  width: 95%;
  margin: 0.5rem;
  display: flex;
  justify-content: space-between;
`;

const Accept = styled.button`
  color: var(--main-black);
  outline: none;
  border: none;
  font-size: 3.01rem;
  background: transparent;
  transition: 0.2s;
  &:hover {
    color: var(--main-green);
    transition: 0.2s;
  }
`;

const Decline = styled.button`
  color: var(--main-black);
  outline: none;
  border: none;
  font-size: 3.01rem;
  background: transparent;
  transition: 0.2s;
  &:hover {
    color: var(--main-red);
    transition: 0.2s;
  }
`;

function PendingChats({ _id, inviter, reciever, status, socketRef, setLocalStatus }: IPendingChats) {
  const [image, setImage] = React.useState<string>("");
  const emitFriendRequest = async () => {
    socketRef?.emit("friend_request");
  };

  const getUserImage = async (name: string) => {
    try {
      const res = await axios.get(`${requestUrl}/users/${name}`);
      const userAvatar = res.data.user.userAvatar;
      if (!userAvatar) {
        setImage("");
        return false;
      }
      const requestString = `${requestUrl}/${userAvatar}`;
      setImage(requestString);
      return true;
    } catch (error) {
      return false;
    }
  };

  const updateInviteStatus = async (param: string) => {
    try {
      setLocalStatus(param);
      await axios.put(`${requestUrl}/invites`, {
        id: _id,
        status: param,
      });
      emitFriendRequest();
      setLocalStatus("");

      return true;
    } catch (error) {
      return false;
    }
  };

  const createChatRoom = async () => {
    try {
      setLocalStatus("accepted");
      await axios.put(`${requestUrl}/chat-room`, {
        id: _id,
        user1: inviter,
        user2: reciever,
      });
      emitFriendRequest();
      setLocalStatus("");
    } catch (error) {
      return false;
    }
  };

  React.useEffect(() => {
    getUserImage(inviter);
  }, []);

  return (
    <div style={{ width: "100%" }}>
      {status === "recieved" && (
        <Contacts className="contacts">
          <div className="user_info flex">
            {image ? <img src={image} alt="user_avatar" /> : <FaUserCircle />}
            <h1 className="flex">{inviter}</h1>
          </div>
          <div className="invite_buttons">
            <Accept
              onClick={() => {
                createChatRoom();
              }}
              className="accept flex"
            >
              <AiFillCheckCircle
                className={css`
                  width: 2rem;
                  height: 2rem;
                `}
              />
            </Accept>
            <Decline onClick={() => updateInviteStatus("declined")} className="decline flex">
              <AiFillCloseCircle
                className={css`
                  width: 2rem;
                  height: 2rem;
                `}
              />
            </Decline>
          </div>
        </Contacts>
      )}
    </div>
  );
}

export default PendingChats;
