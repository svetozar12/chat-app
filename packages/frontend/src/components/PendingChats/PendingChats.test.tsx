import PendingChats from "../PendingChats/PendingChats";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

const socketRef: any = jest.fn();
const setLocalStatus = jest.fn();

const setupRender = () => {
  const component = render(
    <PendingChats
      _id="61c4957b735b579e5442dfe8"
      inviter="ivan"
      reciever="gosho"
      socketRef={socketRef}
      setLocalStatus={setLocalStatus}
      status="recieved"
    />,
  );
  return component;
};

describe("Render connected React-redux page", () => {
  beforeEach(() => {
    setupRender();
  });
  it("should create snapshot for <PendingChats/>", () => {
    expect(
      renderer
        .create(
          <PendingChats
            _id="61c4957b735b579e5442dfe8"
            inviter="ivan"
            reciever="gosho"
            socketRef={socketRef}
            setLocalStatus={setLocalStatus}
            status="recieved"
          />,
        )
        .toJSON(),
    ).toMatchSnapshot();
  });

  it("should render <PendingChats/>", () => {
    const renderedComponent = screen.getByText("ivan");
    expect(renderedComponent).toBeInTheDocument();
  });
});
