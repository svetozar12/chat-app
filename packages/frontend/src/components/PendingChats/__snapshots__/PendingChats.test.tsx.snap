// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render connected React-redux page should create snapshot for <PendingChats/> 1`] = `
<div
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="contacts css-72cnkr"
  >
    <div
      className="user_info flex"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 496 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
        />
      </svg>
      <h1
        className="flex"
      >
        ivan
      </h1>
    </div>
    <div
      className="invite_buttons"
    >
      <button
        className="css-1vgkxxq css-sokyuw"
        onClick={[Function]}
      >
        <svg
          className="css-1q3bwnf"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 0 1-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
          />
        </svg>
      </button>
      <button
        className="css-1vgkxxq css-1cdmtdh"
        onClick={[Function]}
      >
        <svg
          className="css-1q3bwnf"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
